app/src/composables/use-mutex.ts

```diff
@@ -0,0 +1,59 @@
+import { sleep } from '@directus/sdk';
+
+const MutexKey = ['auth_refresh'] as const;
+
+type MutexKey = (typeof MutexKey)[number];
+
+const timeout = 500;
+const maxRetries = 10;
+
+export function useMutex(key: MutexKey, expiresMs: number) {
+	const internalKey = `directus-mutex-${key}`;
+	const useWebLock = !!navigator.locks;
+
+	async function acquireMutex(callback: (lock?: Lock | null) => Promise<any>): Promise<any> {
+		if (useWebLock) {
+			return navigator.locks.request(internalKey, callback);
+		}
+
+		// fall back to localstorage when navigator.locks is not available
+		return localStorageLock(callback);
+	}
+
+	async function localStorageLock(callback: (lock?: Lock | null) => Promise<any>) {
+		let retries = 0;
+		let hasAcquiredMutex = false;
+
+		try {
+			do {
+				// Attempt to prevent concurrent mutex acquiring across browser windows/tabs
+				await sleep(Math.random() * timeout);
+
+				const mutex = localStorage.getItem(internalKey);
+
+				if (!mutex || Number(mutex) > Date.now() + expiresMs) {
+					// set lock
+					localStorage.setItem(internalKey, String(Date.now() + expiresMs));
+					hasAcquiredMutex = true;
+
+					// do logic
+					await callback(null);
+
+					break;
+				}
+
+				retries += 1;
+			} while (retries < maxRetries);
+			// throw error when hitting max retries?
```

Comment:
For our current use case of `useMutex` it probably makes sense to ignore that rather unexpected case (i.e. simply skip refresh).
But I think it would indeed be cleaner to throw an error here and catch it on the caller side.
I think this could be the same if we decide to add a timeout when using Web Lock and that timeout is reached.